Security policy (#8006)
gfx-rs/wgpu: SECURITY.md

# WGPU Security Policy

> This document describes what is considered a security vulnerability in WGPU and

Reviewer: Isn't the name of the project generally written “`wgpu`”, not “WGPU”? Not that that isn’t awkward sometimes, but this should be consistent with the rest of the project documentation.

(Arguably, though, the name `wgpu` should be interpreted as referring to the Rust crate `wgpu`, which I assume _isn't_ used by Firefox for its WebGPU implementation?)

Author: It looks like we don't use either one consistently. The README uses "wgpu", but the GOVERNANCE and CONTRIBUTING documents use "WGPU".

I'm happy to change this document to whichever is preferred.

Reviewer 2: > The README uses "wgpu", but the GOVERNANCE and CONTRIBUTING documents use "WGPU".

I think this is a byproduct of whoever wrote it, I wrote the readme (mainly) and Erich wrote the contrib docs. I have always styled it as "wgpu" no matter what, as I think the capital versions is awkward to me. Lets defer this issue as we can bikeshed it for a while and discuss styling in the next meeting.

how vulnerabilities should be reported.

## Vulnerability Definition

WebGPU introduces a different threat model than is sometimes applied to
GPU-related software. Unlike typical gaming or high-performance computing
applications, where the software accessing GPU APIs is proprietary or
obtained from a trusted developer, WebGPU makes GPU APIs available to
arbitrary web applications. In the threat model of the web, malicious
content should not be able to use the GPU APIs to access data or interfaces

> outside the intended scope for interaction with web content.

Reviewer: I think this paragraph could benefit from spelling out its point more explicitly, because the following text could be taken as _limiting_ the scope of what is a vulnerability (“we mostly care about JavaScript WebGPU”), rather than _expanding_ it (“we care about protection from misuse, not just doing the wrong thing in response to valid-according-to-the-spec input”). So, how about saying something like:

```suggestion
outside the intended scope for interaction with web content.
Therefore, `wgpu` seeks to prevent undefined behavior and data
leaks even when its API is misused, and failures to do so may be
considered vulnerabilities.
(This is also in accordance with the Rust principle of safe vs. unsafe code,
since the `wgpu` library exposes a safe API.)
```

Author: This seems reasonable to me, but I'll wait for some more input before changing it.

The scope as I wrote it was intentionally narrow, because I didn't want to sign the project up for more work than necessary. But that concern aside, I agree that UB or data leaks reachable from the API ought to be considered a vulnerability.

Reviewer 2: I think kpreids suggestion is fine and we should adopt it.

Reviewer 3: What about the experimental APIs? They explicitly call out they may have UB and should be used with caution. Would these still get vulnerability reports? If so what would happen?

Author: @Vecvec I don't think the intent is to make a firm rule that UB in an experimental API is a vulnerability. It's more a statement that the sentiment of the project is that UB is always bad, so it would always be a bug if you can get UB by misusing the API. (Arguably that should be a given for a safe API, but in this context it's worth stating a position on arguable points.)

It doesn't show up in this comment thread, but the paragraph immediately following this should help to provide some clarity that experimental APIs are not held to the same standard.

Author: Sorry, it's not the paragraph immediately following. It's the one a bit later starting "The WGPU Rust API offers some functionality, both supported and experimental...".

Author: To give some examples, I'd say that a simple RCE of the form "I give wgpu this carefully crafted buffer and it executes part of it with no sandbox" is a vulnerability, even in an experimental API. RCE of the form "if you have X and Y in memory for use as ROP gadgets and I can guess the stack canary and it's a particular phase of the moon" might not be a vulnerability in an experimental API. A DoS using a core API might be a vulnerability, a DoS using an experimental API is unlikely to be considered a vulnerability.

But fundamentally, it's often not possible to anticipate what will go wrong; the goal here is just to have some security policy in place to make it clear that we do care about security, and to explain how to report vulnerabilities. Severity (on a scale down to "not a vulnerability at all") has to be assessed on a case-by-case basis.

Reviewer 3: >It doesn't show up in this comment thread, but the paragraph immediately following this should help to provide some clarity that experimental APIs are not held to the same standard.

Ah yes, I see it now.

While the WGPU maintainers have discretion in assigning a severity to

> individual vulnerabilities, it is generally considered a security vulnerability

Reviewer: ```suggestion
individual vulnerabilities, it is generally considered a high-severity vulnerability
```

> in WGPU if Javascript or WebAssembly code running with privileges of ordinary

Reviewer: ```suggestion
in WGPU if JavaScript or WebAssembly code running with privileges of ordinary
```

> web content is able to:

Reviewer: ```suggestion
web content, in a web browser that is using `wgpu` to provide the WebGPU API to that content, is able to:
```

Many users think only about `wgpu` as “abstraction layer for their graphics code” not as “implements WebGPU for web browsers”. I think this would benefit from introducing further explanation that clarifies that  this entire topic is irrelevant to uses of `wgpu` _inside_ WebAssembly programs.

- Access data associated with native applications other than the user agent,
  or associated with other web origins.
- Escape the applicable sandbox and run arbitrary code or call arbitrary system
  APIs on the user agent host.
- Consume system resources to the point that it is difficult to recover
  (e.g. by closing the web page).

The WGPU Rust API offers some functionality, both supported and experimental,
that is not part of the WebGPU standard and is not made available in Javascript
environments using WGPU. Associated vulnerabilities may be assigned lower

> severity than vulnerabilities that apply to a Javascript WebGPU implementation

Reviewer: ```suggestion
severity than vulnerabilities that apply to a WebGPU implementation exposed to JavaScript
```

security vulnerabilities found shortly after the initial release of a major
version may also be provided for the previous major release.

Mozilla provides security support for versions of WGPU used in [current
versions of Firefox](https://whattrainisitnow.com/).

The version of WGPU that is active can be found in the Firefox repositories:

- [release](https://github.com/mozilla-firefox/firefox/blob/release/gfx/wgpu_bindings/Cargo.toml),
- [beta](https://github.com/mozilla-firefox/firefox/blob/beta/gfx/wgpu_bindings/Cargo.toml), and
- [nightly](https://github.com/mozilla-firefox/firefox/blob/main/gfx/wgpu_bindings/Cargo.toml),

We welcome reports of security vulnerabilities in any of these released
versions or in the latest code on the `trunk` branch.

## Reporting a Vulnerability

Although not all vulnerabilities in WGPU will affect Firefox, Mozilla accepts
all vulnerability reports for WGPU and directs them appropriately. Additionally,
Mozilla serves as the CVE numbering authority for the WGPU project.

To report a security problem with WGPU, create a bug in Mozilla's Bugzilla
instance in the [Core ::

> Graphics :: WebGPU](https://bugzilla.mozilla.org/enter_bug.cgi?product=Core&component=Graphics%3A+WebGPU)

Reviewer 4: suggestion: This is a good link, but per the guidance below, I think it'd be even better to bake the security-relevant groups (`core-security`, `gfx-core-security`) into the link, too, so they're ticked by default, i.e.:

https://bugzilla.mozilla.org/enter_bug.cgi?product=Core&component=Graphics%3A+WebGPU&groups=core-security&groups=gfx-core-security

One less potential ~~paper cut~~ big problem to run into. 🙂

Author: I don't know what we might do based on the URL behind the scenes, but the checkbox for the graphics security bug group only shows up if you're in the group. I assume the desire is to have a single ingress point for reports, or maybe that's just a default behavior of groups in bugzilla without any intention for the graphics security group in particular.

Reviewer 4: Just tried running through my proposed link through a private window. Looks like it doesn't check the box like it does when using an employee account. Nuts! It doesn't seem to hurt anything (the end result was [bug 1979455](https://bugzilla.mozilla.org/show_bug.cgi?id=1979455)), but it also doesn't actually solve anything for the audience I was hoping for, either. 🤔

I guess I'd suggest using the URL anyway, since at least it'd save a step for folks who already have group membership? Not firmly set on that, though. 🤷🏻‍♂️

Reviewer 2: If there's no harm, lets do it.

Author: I think I meant to reply to this before, but didn't. Having them both specified in the URL seems like it results in correct labeling of none of the bugs -- if the user doesn't have permission, then the bug is labeled core-security only; if the user does have permission, the bug gets both labels and still needs to be removed from core-security. Although I guess we can have a bot do that. Okay, I'll add gfx-sec to the URL.

> component.

Reviewer 4: style: This seems oddly reflowed. 🤔 Did you use Prettier on this?

Author: It was how my editor reflowed it, prettier doesn't seem to have an opinion. I put the entire link on its own line since that seems more readable.

> **IMPORTANT: For security issues, please make sure that you check the box

Reviewer 4: suggestion: I think we should also include the `gfx-core-security` group checkbox, whose current text of which is `Security-Sensitive Graphics Bug`.

Reviewer 4: Given [discussion](https://github.com/gfx-rs/wgpu/pull/8006#discussion_r2232000689), if this note is added, it'd need to be acknowledged that somebody might not see it if they're not part of the Graphics security group.
